generator/src/main/java/com/graphicsfuzz/generator/tool/CustomMutatorSever.java

```diff
@@ -0,0 +1,129 @@
+import com.graphicsfuzz.common.util.ParseHelper;
+import com.graphicsfuzz.common.util.ParseTimeoutException;
+import com.graphicsfuzz.common.util.RandomWrapper;
+import com.graphicsfuzz.common.util.ShaderKind;
+import com.graphicsfuzz.generator.fuzzer.FuzzedIntoACornerException;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.PrintStream;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+
+/**
+ * Server that can accept shaders from a libFuzzer custom mutator and send back a mutated shader.
+ */
+public class CustomMutatorSever {
+  public static void main(String[] args) {
+    try {
+      // TODO(metzman): If we don't switch from TCP, allow this to be configurable.
+      runServer(8666);
+    } catch (IOException | ParseTimeoutException | InterruptedException exception) {
+      exception.printStackTrace();
+      System.exit(1);
+    }
+  }
+
+  private static void runServer(int port)
+      throws IOException, ParseTimeoutException, InterruptedException {
+    byte[] headerBuff = new byte[28];
+    ServerSocket serverSocket = new ServerSocket(port);
+    Socket socket = serverSocket.accept();
+    InputStream inputStream = socket.getInputStream();
+    OutputStream outputStream = socket.getOutputStream();
+    while (true) {
+      // TODO(metzman): Figure out a better way to handle waiting for the header to arrive.
+      while (inputStream.available() < headerBuff.length) {
+        ;
+      }
+      inputStream.read(headerBuff, 0, headerBuff.length);
+      ByteBuffer headerByteBuffer = ByteBuffer.wrap(headerBuff);
+      headerByteBuffer.order(ByteOrder.LITTLE_ENDIAN);
+
+      long size1 = headerByteBuffer.getLong();
+      // Support two sizes for now so that one message format can be used for both CustomMutate and
+      // CustomCrossOver.
+      long size2 = headerByteBuffer.getLong();
+      assert size2 == 0;
+
+      // Java won't allow us to create an array with a "long" size. Therefore, we must convert it to
+      // int before creating the array. This is probably a non-issue because libFuzzer is unlikely
+      // to ever give us a shader larger than Integer.MAX_VALUE (e.g. ~2G).
+      assert size1 <= Integer.MAX_VALUE && size2 <= Integer.MAX_VALUE;
+
+      // GraphicsFuzz can't do anything with this unfortunately.
+      long maxOutSize = headerByteBuffer.getLong();
+
+      int libfuzzerSeed = headerByteBuffer.getInt();
+
+      // Read the Shader.
+      byte[] inputShaderBuff = new byte[((int) size1) + ((int) size2)];
+      socket.getInputStream().read(inputShaderBuff, 0, inputShaderBuff.length);
+      String inputShader = new String(inputShaderBuff);
+
+      try {
+        // Now mutate the shader.
+        TranslationUnit tu = ParseHelper.parse(inputShader, ShaderKind.FRAGMENT);
```

Comment:
Please document the assumption that the shader is a fragment shader.  I guess from a libFuzzer perspective the shader kind is unimportant?  (And indeed we then usually use ShaderKind.FRAGMENT as a default.)

Reply:
It does sort of matter for libFuzzer in this particular case. I changed things so libFuzzer can instruct GraphicsFuzz to use a fragment or vertex.